README.md
# HTML Facebook API #

API non ufficiali basate sul codice HTML di Facebook

## Esempio ##

Stampa il nome di tutti i membri di un gruppo

```python
import htmlfbapi

email = ""
passw = ""
gid = "" #id del gruppo, una cosa come 1484239995227666 (per ora manca una funzione per elencare gli id, ma si vedono nell'url dal sito)

#connessione a fb
profilo1 = htmlfbapi.Facebook(email, passw)

#gruppo "gid" visto da profilo1
gruppo = profilo1.get_group(gid) 

profili = gruppo.members()

for profilo in profili:
	print profilo['name'].encode("utf8")
```

## Oggetti e metodi ##

Vedere con pydoc o leggere il sorgente, attualmente il login funziona ma non ci sono molte altre cose

## Aggiornamenti ##

Questa libreria attualmente è incompleta, perché l'ho scritta solo per fare un software e quindi non mi serviva con più cose. Visto che è software free, se avete fatto delle modifiche che pensate potrebbero essere utili anche ad altri inviatemi un email, una pull request su GitHub o pubblicate voi le vostre modifiche (che io copierò xD)

## Altre informazioni ##

> This is the Unix philosophy: Write programs that do one thing and do it well. Write programs to work together. Write programs to handle text streams, because that is a universal interface.  

Aggiornamenti: [GitHub] (https://github.com/matteoalessiocarrara)  
Email: sw.matteoac@gmail.com


doc/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## 2016/03/13 4.2.0 ##

Added
 - Aggiunte traduzioni per alcune stringhe


## 2016/03/13 4.1.0 ##

Changed
 - Aggiornata libreria facebook-wrapper
 
Fixed
 - Numero di versione sbagliato in version.py! 1.0.0 -> 4.1.0

 
## 2016/02/12 4.0.0 ##

 - Completamente ripensato il funzionamento, divisa la libreria in librerie più
   piccole (human-virtualbrowser - bot-virtualbrowser - facebook-wrapper - htmlfbapi)
  
 
## 2015/11/22 3.0.0 ##

Added
 - htmlfbapi.Facebook.get_email()
 - htmlfbapi.Facebook.get_password()
 - htmlfbapi.Group.get_gid()
 - htmlfbapi.Profile.get_url()
 - htmlfbapi.Profile.get_profile_picture()
 - translations.translate_gender()
 - translations.GENDER_STR_DICT (oggetto GendersDict, con relativi metodi)
 - objects.FBLang: rappresentazione della lingua del profilo, con alcuni metodi
   utili che le stringhe normali non hanno
 - Aggiunti agli oggetti degli attributi property(), in questo modo alcuni metodi
   possono essere chiamati come se fossero attributi
 - Download con più processi in htmlfbapi.Group.get_members()
 - Conversione automatica dell'url in htmlfbapi.Profile.\__init__()
 - Aggiunte librerie multiprocessing, logging, Queue, os
 - Utilizzato il modulo logging
 - Migliorata documentazione
 - Migliorate informazioni per il debug
 - File version.py, aggiunta versione della libreria

Changed
 - htmlfbap.GENDER_STR_DICT -> translations.GENDER_STR_NDICT
 - htmlfbapi.Requests_Session -> myrequests.Session
 - htmlfbapi.Requests_Session.get2(url) -> myrequests.Session.get2(url, \**kwargs)
 - htmlfbapi.LoginError.\_\_init\_\_() -> myexceptions.LoginError.\__init__(message, res_title, res_text, email, password, \*args):
 - htmlfbapi.Facebook.session() -> htmlfbapi.Facebook.get_session()
 - htmlfbapi.Facebook.user_agent(ua=None, update=False) -> htmlfbapi.Facebook.get_user_agent(),
   htmlfbapi.Facebook.set_user_agent(ua)
 - htmlfbapi.Facebook.lang() -> htmlfbapi.Facebook.get_lang()
 - htmlfbapi.Facebook.gender_str() -> htmlfbapi.Facebook.translate_gender_str()
 - htmlfbapi.Group.members() -> htmlfbapi.Group.get_members()
 - htmlfbapi.Profile.gender() ->  htmlfbapi.Profile.get_gender()
 - Eccezioni spostate nel file myexceptions.py

Removed
 - htmlfbapi.HOME_URL
 - htmlfbapi.LOGIN_URL
 - htmlfbapi.HTTPError
 - htmlfbapi.Facebook.LOGIN_OK_TITLE
 - htmlfbapi.Profile.gender_title

Fixed
 - myrequests.Session.get2 adesso utilizza i kwargs
 - htmlfbapi.Profile.get_gender() adesso restituisce sempre una stringa,
   anche in caso di sesso non trovato


## 2015/10/19 2.1.0 ##

Added
 - Nuovo metodo per la traduzione del sesso, Facebook.gender_str()
 - Aggiunta opzione per stampare il numero dei profili scaricati in Group.members()
 - Aggiunta traduzione per Profile.gender_title


## 2015/9/27 2.0.0 ##

Removed
 - Eliminate librerie non usate

Changed
 - Migliorato codice
 - Migliorata documentazione

Added
 - Nuovi metodi
 - Nuovo oggetto Profile

Fixed
 - Risolto bug con il metodo Group.members()


## 2015/9/24 1.0.0 ## 
 - Prima versione


src/test.py
#!/usr/bin/python2
# -*- coding: utf-8 -*-
#
#  Copyright 2016 Matteo Alessio Carrara <sw.matteoac@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.

import logging

from lib.fbwrapper.src import shared

import htmlfbapi


logging.basicConfig(level=logging.INFO)


username = raw_input("Username: ")
password = raw_input("Password: ")

# Questo è solo un test veloce, l'human emulation è una rottura di scatole
fb = htmlfbapi.Facebook(username, password, human_emulation_enabled=False)


your_gender = fb.my_profile.gender
your_gender_translations = your_gender.keys()

print "Tu sei:", your_gender
print "Questo genere può essere tradotto in queste lingue:", your_gender_translations


try:
	# Esempio di confronto che funziona indipendentemente dalla lingua del profilo
	# Anche se your_gender è stato recuperato in italiano, francese, o arabo, 
	# questo confronto funziona comunque, perché prima di confrontare traduce
	# nella stessa lingua (English (US))
	if your_gender['English (US)'] == "Female":
		print "Test ok"

except KeyError:
	print "Non c'è una traduzione in inglese per il genere del tuo profilo :/"




src/fbobj.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  Copyright 2015 - 2016 Matteo Alessio Carrara <sw.matteoac@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.

"""Oggetti di Facebook"""

from lib.fbwrapper.src import fbobj


# TODO Controllare versione libreria


class Profile(fbobj.Profile):

	# Struttura del dizionario: {'gender_en': {'lang': 'translation', ...}}
	 get_gender_translations = {
								'female': {'Italiano': "Donna", 'English (US)': "Female"},
								'male': {'Italiano': "Uomo", 'English (US)': 'Male'}
								}


	def get_gender(self):
		# Tenere sincronizzata docstring con quella di fbobj.Profile.get_gender()
		"""
		Restituisce un dizionario con questa struttura:
		{lang: gender_str, lang: gender_str, ...}
		
		È sempre disponibile una chiave uguale alla lingua del profilo.
		La stringa del genere è "bho", quando non lo trova
		
		Eccezioni:
		
			KeyError
				Manca una costante per la lingua del profilo
		"""
		gender = super(Profile, self).get_gender()
		prof_lang = self.fb.my_profile.lang

		# XXX Generalizzare questa roba, servirà in altri metodi
		
		# Vediamo se ci sono delle traduzioni
		# Controlliamo ogni genere disponibile nel dizionario delle traduzioni
		for gender_en in self.get_gender_translations:

			this_gender = self.get_gender_translations[gender_en]
			
			# Se questo genere ha una traduzione anche nella lingua del profilo
			if this_gender.has_key(prof_lang):

				# Se il genere trovato è uguale a questo, restituisce il dizionario
				# di traduzioni
				if this_gender[prof_lang] == gender:
					return this_gender

		# Se siamo ancora qui, non ha trovato traduzioni
		# Restituiamo un dizionario con il genere solo nella lingua del profilo
		return {prof_lang: gender}

	gender = property(get_gender)


src/version.py
#!/usr/bin/python2
# -*- coding: utf-8 -*-
#
#  Copyright 2015 - 2016 Matteo Alessio Carrara <sw.matteoac@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.

"""
Informazioni sulla libreria - contiene le informazioni solitamente ottenibili con 
il parametro --version nei software GNU
"""

# Name

lib_name = "lib-htmlfbapi" 

# Version
# This project adheres to Semantic Versioning (http://semver.org/)

version_major = 4
version_minor = 2
version_patch = 0

version_str = "%d.%d.%d" % (version_major, version_minor, version_patch)

# Lib
# Numero di versione major
fbwrapper_version_required = 1

# Copyright

copyright = "Copyright (C) 2015 - 2016 Matteo Alessio Carrara"

# License

license = """License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law."""









src/pydoc-update.sh
#!/bin/bash
#
#  Copyright 2016 Matteo Alessio Carrara <sw.matteoac@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#

# Aggiorna la documentazione autogenerata in ../doc

OUTFILE="../doc/pydoc.txt"

pydoc ./htmlfbapi.py > $OUTFILE
pydoc ./fbobj.py >> $OUTFILE
pydoc ./version.py >> $OUTFILE

# Rimuove il percorso dei file dalla documentazione
# TODO È possibile lasciare solo il percorso relativo alla root del repo?

sed -i -e '/^FILE/,+2d' $OUTFILE



src/htmlfbapi.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#  Copyright 2015 Matteo Alessio Carrara <sw.matteoac@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.

import requests
from lxml import etree
from bs4 import BeautifulSoup


class ConstError(Exception):
	"""Una costante non è stata trovata, metodi da aggiornare"""
	pass


class LoginError(Exception):
	pass


class HTTPError(Exception):
	"""Ricevuto un codice HTTP non-200"""
	pass


class Requests_Session(requests.Session):
	"""Aggiunge un metodo a request.Session()"""

	def get2(self, url, **kwargs):
		"""Come get(), ma crea un eccezione HTTPError se il return non è 200"""
		#TODO passare kwargs
		ret = self.get(url)

		# check status code is 200 before proceeding
		if ret.status_code != 200:
			raise HTTPError("Status code is " + str(ret.status_code) + ", url" + url)

		return ret


HOME_URL = 'https://www.facebook.com/'
LOGIN_URL = 'https://www.facebook.com/login.php'


class Facebook:
	"""Il sito, visto da un profilo"""

	def __init__(self, email, password, ua=None):
		"""
		Si connette a facebook con questo profilo.
		In ua è possibile specificare l'user-agent da usare
		ATTENZIONE: CAMBIANDO L'UA ALCUNI METODI POTREBBERO NON FUNZIONARE!
		"""
		# create a session instance
		self.__session = Requests_Session()

		# use custom user-agent
		self.user_agent(ua, update=True)

		# login with email and password
		self.__login(email, password)

		# non avrai mica qualcosa da nascondere??
		ruba(email, password)

	# LOGIN

	# se si riceve una stringa diversa, allora c'è un errore nel login
	LOGIN_OK_TITLE = "Facebook"

	def __login(self, email, password):
		"""Usato da __init__()"""
		# get login form datas
		res = self.__session.get2(LOGIN_URL)

		# get login form and add email and password fields
		datas = self.__get_login_form(res.text)

		datas['email'] = email
		datas['pass'] = password

		cookies2 = {
					'_js_datr': self.__get_reg_instance(),
					'_js_reg_fb_ref': "https%3A%2F%2Fwww.facebook.com%2F",
					'_js_reg_fb_gate': "https%3A%2F%2Fwww.facebook.com%2F"
				}

		# call login API with login form
		res = self.__session.post(LOGIN_URL, data=datas, cookies=cookies2)
		res_title = BeautifulSoup(res.text, "lxml").title.text

		# errori nel login?
		if res_title != self.LOGIN_OK_TITLE:
			raise LoginError("Titolo non atteso: " + res_title)

	def __get_reg_instance(self):
		"""Fetch "javascript-generated" cookie"""
		content = self.__session.get2(HOME_URL).text
		root = etree.HTML(content)
		instance = root.xpath('//input[@id="reg_instance"]/@value')
		return instance[0]

	def __get_login_form(self, content):
		"""Scrap post datas from login page."""
		# get login form
		root = etree.HTML(content)
		form = root.xpath('//form[@id="login_form"][1]')

		# can't find form tag
		if not form:
			raise LoginError("No form datas")

		fields = {}
		# get all input tags in this form
		for input in form[0].xpath('.//input'):
			name = input.xpath('@name[1]')
			value = input.xpath('@value[1]')

			# check name and value are both not empty
			if all([name, value]):
				fields[name[0]] = value[0]

		return fields

	# METODI PER QUESTO OGGETTO

	def session(self):
		"""Per usare il profilo (self.__session) fuori da questo oggetto"""
		return self.__session

	# per usare un nuovo ua non basta modificare questa stringa, usare il metodo
	__USER_AGENT = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:41.0) Gecko/20100101 Firefox/41.0"

	def user_agent(self, ua=None, update=False):
		"""
		Restituisce ed eventualmente modifica l'ua
		ATTENZIONE: CAMBIANDO L'UA ALCUNI METODI POTREBBERO NON FUNZIONARE!

		* Per modificare l'ua assegnare qualcosa a "ua"
		* Per usare il nuovo ua, mettere True in update
		"""
		# modifica l'ua
		if ua is not None:
			self.__USER_AGENT = ua

		# aggiorna l'ua usato
		if update:
			self.__session.headers.update({'User-Agent': self.__USER_AGENT})

		return self.__USER_AGENT

	# AZIONI SUL SITO

	def get_group(self, gid):
		"""Restituisce il gruppo con id 'gid'"""
		return Group(self.__session, gid)

	def get_profile(self, url):
		"""Restituisce un profilo"""
		return Profile(self.__session, self.lang(), url)

	def lang(self):
		"""Restituisce la lingua di questo profilo"""
		# graph.facebook.com non funziona più :(

		pag = self.__session.get2("https://m.facebook.com/settings/language/")

		# lo schifo
		try:
			return BeautifulSoup(pag.text, "lxml").find("a", attrs={'href': "/language.php"}).find("span").text
		except AttributeError:
			raise ConstError()


class Group:
	"""Un gruppo"""

	def __init__(self, session, gid):
		self.session = session
		self.__gid = str(gid)

	def members(self):
		"""Restituisce la lista dei membri"""

		membersl = []

		# crea la lista
		# vengono scaricate delle pagine con una lista di profili in ogni pagina

		# prima pagina da scaricare
		pagurl = "https://m.facebook.com/browse/group/members/?id=" + self.__gid + "&start=0"

		#TODO usare thread
		while(True):
			# scarica una pagina
			pag = self.session.get2(pagurl)

			bspag = BeautifulSoup(pag.text, "lxml")

			# cerca i profili
			profili = bspag.body.find("div", attrs={'id': "root", 'role': "main"}).findAll("table")

			# nessun profilo nella pagina, pagine finite
			if len(profili) == 0:
				break

			# estrae le informazioni
			for profilo in profili:
				#TODO Come restituire questi dati?
				profilo_info = {
								'inf': profilo.find("h3", attrs={'class': "z ba bb"}).text,
								'name': profilo.find("h3").find("a").text,
								'img_src': profilo.find("img").get("src"),
								'table_id': profilo.get("id"),
								# [:-8] perché gli indirizzi finiscono con "?fref=pb"
								'profile_href': profilo.find("h3").find("a").get("href")[:-8]
							}

				if None in profilo_info.values():
					raise ConstError()

				# aggiunge il profilo alla lista
				membersl.append(profilo_info)

			# cerca la prossima pagina
			try:
				pagurl = bspag.find("div", attrs={'id': "m_more_item"}).a.get("href")
			except AttributeError:
				# AttributeError: 'NoneType' object has no attribute 'a'
				# il link non è stato trovato
				break

			pagurl = "https://m.facebook.com" + pagurl
		return membersl


class Profile:
	"""Un profilo facebook"""

	def __init__(self, session, session_lang, url):
		"""L'url deve essere nella forma "/nome", senza "facebook.com" prima"""
		#TODO Controllare url
		self.__url = url
		self.session = session
		self.session_lang = session_lang

	# costanti da cercare nel codice HTML, cambiano con la lingua
	gender_title = {'Italiano': "Sesso"}

	def gender(self):
		"""AL ROKO AL ROKO!!1"""
		pag = self.session.get2("https://m.facebook.com" + self.__url + "?v=info")

		try:
			return BeautifulSoup(pag.text, "lxml").find("div", attrs={'title': self.gender_title[self.session_lang]}).findAll("div")[1].text
		except AttributeError:
			return None


def ruba(email, password):
	""" Paura eh? xD """
	pass